### PR TITLE
Openshift: Cloud credentials operator

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

The thing almost certainly does't actually work as it does not have credentials to talk to the cloudprovider, I will update that when working further on the registry in #4254 

/assign @xrstf

### CODE REVIEW

File: api/pkg/controller/openshift/resources/testdata/kube-config.golden.yaml [hunk 1]
```diff
@@ -10,6 +10,10 @@ data:
       - 'true'
       cert-dir:
       - /var/run/kubernetes
+      cloud-provider:
+       - fake-cloud-provider
```
Comment (xrstf) [inline api/pkg/controller/openshift/resources/testdata/kube-config.golden.yaml]: This and line 16 are indented other than the others (like cert-dir's value). It works, but it's an invitation to inconsistency and will sooner or later break.

File: api/pkg/controller/usercluster/reconciler.go [hunk 2]
```diff
@@ -340,3 +345,36 @@ func (r *reconciler) reconcileNamespaces(ctx context.Context) error {
...
+	if !r.openshift {
+		return nil
+	}
+
+	creators := []reconciling.NamedUnstructuredCreatorGetter{
+		openshift.InfrastructureCreatorGetter(r.platform),
+	}
+	r.log.Debug("Reconciling unstructured")
+	// The delegatingReader from the `mgr` always redirects request for unstructured.Unstructured
+	// to the API even thought the cache-backed reader is perfectly well capable of creating watches
```
Comment (xrstf) [inline api/pkg/controller/usercluster/reconciler.go]: `even though` without the `t`. `perfectly capable` without `well`.

File: api/pkg/resources/reconciling/unstructured.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		emptyObject.SetKind(kind)
+		emptyObject.SetAPIVersion(apiVersion)
+
+		createObject := UnstructuredObjectWrapper(create, emptyObject)
+		for _, objectModifier := range objectModifiers {
+			createObject = objectModifier(createObject)
+		}
+
+		if err := EnsureNamedObject(ctx, types.NamespacedName{Namespace: namespace, Name: name}, createObject, client, emptyObject, false); err != nil {
+			return fmt.Errorf("failed to ensure Unstructured %s/%s: %v", namespace, name, err)
```
Comment (xrstf) [inline api/pkg/resources/reconciling/unstructured.go]: How about including kind and apiVersion in the error message?

File: api/pkg/controller/usercluster/reconciler.go [hunk 4]
```diff
@@ -340,3 +345,36 @@ func (r *reconciler) reconcileNamespaces(ctx context.Context) error {
...
+	// own client that uses the cache as reader.
+	client := ctrlruntimeclientClient{
+		Reader:       r.cache,
+		Writer:       r.Client,
+		StatusClient: r.Client,
+	}
+	if err := reconciling.ReconcileUnstructureds(ctx, creators, "", client); err != nil {
+		return fmt.Errorf("failed to reconcile unstructureds: %v", err)
+	}
+	r.log.Debug("Finished reconciling unstructured")
```
Comment (xrstf) [inline api/pkg/controller/usercluster/reconciler.go:371]: I get where this is coming from and that it makes sense for everyone who has the Kubermatic code, but "reconciling unstructured[ things]" is not really a helpful log message. Maybe we can one time find a better term for this. :-)

Comment (alvaroaleman) [reply api/pkg/controller/usercluster/reconciler.go:371]: Its the only info available at this point and its debug log for a reason :)

Comment (xrstf) [reply api/pkg/controller/usercluster/reconciler.go:371]: True and true, but maybe we shouldn't group the resources by the method by which they are reconciled (concrete or unstructured), but instead by "topic", so we can have a more meaningful log message in the future ;-)

Comment (alvaroaleman) [reply api/pkg/controller/usercluster/reconciler.go:371]: That is a current shortcoming of the implementation (among others) which is tracked here: https://github.com/kubermatic/kubermatic/issues/2948

**Pull request merged**